lib/core/src/Cardano/Wallet/DB/Model.hs

```diff
@@ -0,0 +1,235 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: Apache-2.0
+--
+-- An implementation of the wallet database using only pure functions.
+--
+-- These functions and types model the behaviour of the SQLite database backend,
+-- and are used for QuickCheck state machine testing, and the MVar database
+-- backend.
+
+module Cardano.Wallet.DB.Model
+    (
+    -- * Model Types
+      Database (..)
+    , WalletDatabase (..)
+    , emptyDatabase
+    , TxHistory
+    , TxHistoryMap
+    , filterTxHistory
+    -- * Model Operation Types
+    , ModelOp
+    , Err (..)
+    -- * Model database functions
+    , mCleanDB
+    , mCreateWallet
+    , mRemoveWallet
+    , mListWallets
+    , mPutCheckpoint
+    , mReadCheckpoint
+    , mPutWalletMeta
+    , mReadWalletMeta
+    , mPutTxHistory
+    , mReadTxHistory
+    , mPutPrivateKey
+    , mReadPrivateKey
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Model
+    ( Wallet )
+import Cardano.Wallet.Primitive.Types
+    ( Hash
+    , Range (..)
+    , SlotId (..)
+    , SortOrder (..)
+    , Tx
+    , TxMeta (slotId)
+    , WalletMetadata
+    , isWithinRange
+    )
+import Data.List
+    ( sort, sortOn )
+import Data.Map.Strict
+    ( Map )
+import Data.Maybe
+    ( catMaybes, listToMaybe )
+import Data.Ord
+    ( Down (..) )
+import GHC.Generics
+    ( Generic )
+
+import qualified Data.Map.Strict as Map
+
+{-------------------------------------------------------------------------------
+                            Model Database Types
+-------------------------------------------------------------------------------}
+
+-- | Model database, parameterised by the wallet ID type, the wallet AD state
+-- type, the target backend, and the private key type.
+--
+-- Tne type parameters exist so that simpler mock types can be used in place of
+-- actual wallet types.
+data Database wid s t xprv = Database
+    { wallets :: !(Map wid (WalletDatabase s t xprv))
+    -- ^ Wallet-related information.
+    , txs :: (Map (Hash "Tx") (Tx t))
```

Comment:
Why do we need `Tx`s? And why globally? It is not in your [sqlite schema diagram](https://github.com/input-output-hk/cardano-wallet-adr/issues/3#issuecomment-523010666).

Reply:
Ah yes, it's not in that diagram because it only has tables/columns relevant to rollback.
But in the full schema, a Tx is just all rows of tx_in and tx_out that share a txid. Because transactions are immutable, they are not attached to any wallet.